enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,616 @@
+
+### Preface
+
+Every action related to procedure described in this enhancement is 
+performed after system's boot rather than immediately before shutdown. 
+Greenboot's healthchecks, green and red scripts are executed independent of MicroShift's processes.
+Actions related to backup, restore, and data migration may be performed with MicroShift 
+partially running, i.e. only etcd and kube-apiserver are running.
+
+Only one backup of MicroShift data will be stored at a given moment
+due to high probability of devices having limited storage.
+
+### Integration with greenboot
+
+Depending on result of greenboot's healthcheck either "green" (successful boot) or "red" (unsuccessful) scripts are executed before rebooting the system.
+
+MicroShift will integrate with that system to persist an action to perform on next boot:
+- "green": on next boot, before MicroShift starts, make a backup of MicroShift's data
+- "red": on next boot, before MicroShift starts, restore MicroShift's data from backup
+
+Functionality will be implemented by placing in `/etc/greenboot/green.d` and `/etc/greenboot/red.d`
+bash scripts containing with simple logic or, if needed,
+executing commands `microshift greenboot green` and `microshift greenboot red` in case of needing to put more information into file with action for next boot.
+Alternatively, a symlinks to `microshift` binary can be made and MicroShift modified to run specific command depending on content of argv[0] (just like BusyBox).
+
+File containing said information should not be part of MicroShift data directory
+as we don't to be a part of backup.
+
+### Backup and restore of MicroShift data
+
+Decision whether to backup or restore is based on file persisted during previous boot (see "Integration with greenboot").
+When file is read to make a decision (but not during dry run), it shall be removed.
+
+As a result of investigation and aiming for simplicity for initial implementation,
+it was decided that backing up MicroShift's data will be done by leveraging using copy-on-write (CoW) functionality.
+
+CoW is a feature of filesystem (supported by XFS and Btrfs) and it can be used by providing a `--reflink=` param to `cp` option.
+`--reflink=auto` will be used over `--reflink=always` to gracefully fall back to regular copying on filesystems
+not supporting CoW (ext4, ZFS).
+Backup will be done in `/var/lib/microshift.bak` or similar - it needs to be within the same filesystem/mount.
+Only one backup will be kept. Consequent boots will overwrite the backup dir (only if the action is "backup").
+
+End user documentation needs to include guidance on setting up filesystem to fullfil requirements for using copy-on-write (e.g. making sure some filesystem options are not disabled).
+
+### MicroShift version metadata persistence
+
+When MicroShift is up and running healthy, it will persist its own version into a file within data dir, e.g.:
+```
+4.14.0
+```
```

Comment:
Is this enough info to store? Do we want to support differentiating among same z-version, but different commits?
In any case, we should mention this explicitly one way or another.

Reply:
We could store full version string, but how would we use it?
It would be best to not over complicate things.
Also released software is x.y.z, I don't think we need to involve commits

Reply:
We could add a secondary commit ID and use it if the version is the "same". This should also allow us better flexibility when writing tests, etc.

Reply:
Oh, you meant ostree commit, I understood full microshift version that includes o/microshift commit :laughing: 
This is something I thought about previously (might even still exists in older commits of enhancement) but haven't thought about argument to keep it.
This sounds good to me, adds more safety net.
Could you elaborate how this improve flexibility when testing? I see it as another hard link between `ostree commit - microshift version - backup version` so it's another thing that needs to match others

Reply:
We have access to a bunch of version details

```
$ ./_output/bin/microshift version -o yaml
buildDate: "2023-04-21T21:50:28Z"
compiler: gc
gitCommit: 9edcd2699
gitTreeState: dirty
gitVersion: 4.14.0-0.nightly-arm64-2023-04-18-222945-20230421204708-9edcd2699-dirty
goVersion: go1.19.4
major: "4"
minor: "14"
platform: linux/arm64
```

It seems like x.y.z should be enough, though.